### PR TITLE
Implement SSO passcode-based authentication flow

### LINKED ISSUE #226: Support SSO authentication
Some Cloud Foundry deployments make use of single sign on (SSO) authentication using a passcode:

```
   cf login [-a API_URL] [-u USERNAME] [-p PASSWORD] [-o ORG] [-s SPACE] [--sso | --sso-passcode PASSCODE]
   cf login --sso (cf will provide a url to obtain a one-time passcode to login)
   --sso                      Prompt for a one-time passcode to login
   --sso-passcode             One-time passcode
```

The provider currently doesn't support this authentication flow, and also attempts to login multiple times (in the refresh and plan steps, for example).

### PR DESCRIPTION
**Pull request by vixus0**
Fixes #226

This makes the provider compatible with Cloud Foundry distributions which use single sign on (SSO) instead of username/password to authenticate users.

This adds a new provider parameter `sso_passcode` (environment variable `CF_SSO_PASSCODE`) which takes the same value as `cf login --sso-passcode <passcode>`. This also implements persisting and refreshing the access token to disk to avoid the need for repeated login attempts, which are incompatible with the SSO flow.

### CODE REVIEW

**Comment by janosbinder** [comment]
Thanks for your contribution vixus - I am launching the integration builds meanwhile: https://github.com/cloudfoundry-community/terraform-provider-cf/pull/228 Would you be so kind to also update the documentation?

**Comment by vixus0** [comment]
Hi @janosbinder, thanks for the reminder.

**Comment by janosbinder** [comment]
@ArthurHlt could you please have a look on this PR? From my side looks fine, and I have executed the tests against the upstream here https://github.com/cloudfoundry-community/terraform-provider-cf/pull/229

**Comment by ArthurHlt** [comment]
@janosbinder I didn't wrote about it but the platform which was given by @mevansam has changed all theirs password, we can't access anymore. So, automated test over travis are for now skipped, we need to run it locally sadly.

@vixus0 Thanks for your PR, I do understand the interest for passcode but there is something in the code which I'm not superfan. You're writing all tokens in a file called `.cf-tokens`, this can introduce side effect, I would like to never have to write in local filesystem for getting context. Can you remove this part and just lets the part for getting tokens from passcode ? 

I do understand that users will need to get passcode each time they want to perform terraform plan/apply but at least there is no magic with this file and the behaviour will be predictable.

**Comment by vixus0** [comment]
@ArthurHlt yes you're right about the side effects. I also dislike storing creds in a file but without caching the tokens somewhere authentication needs to be repeated after the Terraform state refresh which requires a new passcode. Is there some other way to use the same token between the refresh and plan stages?

The auth code could use a bit of work in general, and I'm happy to take a look at it. For example, why do we build `*Sess` clients in: https://github.com/cloudfoundry-community/terraform-provider-cf/blob/c6d650e9704df75883a9a266206de49843c05ac4/cloudfoundry/managers/session.go#L302-L338

**Comment by ArthurHlt** [comment]
Well, we can find an alternative. We can set a parameter in provider config in addition of param `sso_passcode` to let user set a path to store tokens. If empty this store nothing.
It can be called `store_tokens_path` and like this it removes `CF_TOKENS_FILE` constant and parametrize this behaviour (by default we don't want it).
Does it looks good to you ? If yes, can you change your PR for that ? Note that it will add a parameter and will need to have doc associated with

For your question, this is for creating an uaa client with uaa admin client for example which is not the same as your cloud foundry user. So yes good catch for me, the part you have added should be remove as your passcode is a cloud foundry user.

File: cloudfoundry/managers/session.go [hunk 1]
```diff
@@ -248,7 +311,10 @@ func (s *Session) init(config *configv3.Config, configUaa *configv3.Config, conf
 
 		var accessTokenSess string
 		var refreshTokenSess string
-		if configUaa.UAAOAuthClient() == "cf" {
+		if configSess.SSOPasscode != "" || loadedTokens {
+			accessTokenSess = accessToken
+			refreshTokenSess = refreshToken
+		} else if configUaa.UAAOAuthClient() == "cf" {
```
Comment (ArthurHlt) [inline cloudfoundry/managers/session.go]: this part should be revert as passcode is a cloud foundry user and not an uaa user

**Comment by ArthurHlt** [comment]
I'm merging it and do the modifications. Going for `store_tokens_path` params solution.

**Comment by ArthurHlt** [comment]
You can see implementation here: https://github.com/cloudfoundry-community/terraform-provider-cf/commit/3f60d901c7bf17ef16ca092c7e131078521d4292

**Comment by ArthurHlt** [comment]
One more thing, when you were asking why we don't use refresh tokens. My explanation was partially wrong. I can explain why we create from client id, this is because if you set username and client id, first tokens will be created from username/password against this client id. But you don't want user for uaa api, you want the client id tokens not user. For the first, when using username, yes simply take previous tokens just works (and that's what i've corrected/used when implementing token store)

**Comment by AndreyZykov** [comment]
@vixus0 could you please provide an example on how to use passcode auth?
I've tried different terraform (0.12.24 / 0.11.14) and provider (0.11.0 / 0.12.0) versions and always get the same error: Invalid passcode
`# CF_SSO_PASSCODE=abvcscdrct terraform plan -var-file="dev.tfvars"
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.cloudfoundry_org.org: Refreshing state...
data.cloudfoundry_space.space: Refreshing state...

------------------------------------------------------------------------

Error: Error running plan: 1 error occurred:
	* provider.cloudfoundry: Error when creating clients: Error when authenticate on cf using SSO passcode: Invalid passcode
`

**Comment by vixus0** [comment]
@AndreyZykov sadly I no longer have an environment in which I can test this easily - where are you getting the passcode from? Does the passcode work with `cf login --sso-passcode`?

**Comment by AndreyZykov** [comment]
@vixus0 I have few environments but passcode auth doesn't work on all of them. For example, on public run.pivotal.io (https://login.run.pivotal.io/passcode):
```
cf login -a api.run.pivotal.io --sso-passcode UibZSR9ua3
API endpoint: api.run.pivotal.io

Authenticating...
OK


Targeted org xxxxxxxx

Targeted space xxxxxxxx



API endpoint:   https://api.run.pivotal.io (API version: 3.82.0)
User:           xxxxxxxx
Org:            xxxxxxxx
Space:          xxxxxxxx
```

and the same with terraform (I've used another passcode, because AFAIU it can be used only one time):

```
CF_SSO_PASSCODE="50s6RccLfT" CF_API_URL="https://api.run.pivotal.io" terraform plan -var-file="dev.tfvars"
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.cloudfoundry_org.org: Refreshing state...
data.cloudfoundry_space.space: Refreshing state...


------------------------------------------------------------------------

Error: Error when creating clients: Error when authenticate on cf using SSO passcode: Invalid passcode

  on main.tf line 41, in provider "cloudfoundry":
  41: provider "cloudfoundry" {
```

**Comment by AndreyZykov** [comment]
I created a sample project to check passcode auth with run.pivotal.io:
https://github.com/AndreyZykov/cf-terraform-test

**Comment by janosbinder** [comment]
@AndreyZykov does it help if you enable this before executing terraform? 

```export CF_DEBUG=true

export CF_TRACE=debug.log```

Otherwise you can also add the debugging lines in the code and see what is passed.

**Comment by AndreyZykov** [comment]
@janosbinder I run these two commands but got no debug from terraform:

```
➜ cf-terraform-test git:(master) export CF_DEBUG=true

➜ cf-terraform-test git:(master) export CF_TRACE=debug.log

➜ cf-terraform-test git:(master) CF_SSO_PASSCODE=UxKGwjBz5T terraform plan -var 'org_name=xxxxxx'
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


Error: Error when creating clients: Error when authenticate on cf using SSO passcode: Invalid passcode

  on main.tf line 11, in provider "cloudfoundry":
  11: provider "cloudfoundry" {



➜ cf-terraform-test git:(master) ls -la
total 24
drwxr-xr-x   7 azykov  staff  224 Apr 28 15:11 .
drwxr-xr-x  22 azykov  staff  704 Apr 28 13:51 ..
drwxr-xr-x  15 azykov  staff  480 Apr 28 15:11 .git
-rw-r--r--   1 azykov  staff   12 Apr 28 14:09 .gitignore
drwxr-xr-x   3 azykov  staff   96 Apr 28 14:05 .terraform
-rw-r--r--   1 azykov  staff  262 Apr 28 14:13 Readme.md
-rw-r--r--   1 azykov  staff  410 Apr 28 14:03 main.tf
```

where exactly I can add debugging lines to find what it passed?

**Comment by janosbinder** [comment]
Hm interesting that debug.log is not generated.

Let us try something different. Can you add a line to provider “cloudfoundry” {
sso_passcode = “mysecretpasscode”
and see what happens if you define it in the tf file and not as a variable?

**Comment by nickromney** [comment]
@AndreyZykov @janosbinder 

I found that I needed to support sso_passcode, I needed to set:

- `sso_passcode` (a one-time token)
- `skip_ssl_validation` = true
- `store_tokens_path` = /path/to/local/file

However, I needed to permit other use cases of user and password, so my provider block looks like:

```
provider cloudfoundry {
  api_url             = var.cloudfoundry_api_url
  user                = var.cloudfoundry_user != "" ? var.cloudfoundry_user : null
  password            = var.cloudfoundry_password != "" ? var.cloudfoundry_password : null
  skip_ssl_validation = var.cloudfoundry_skip_ssl_validation != "" ? var.cloudfoundry_skip_ssl_validation : false
  sso_passcode        = var.cloudfoundry_sso_passcode != "" ? var.cloudfoundry_sso_passcode : null
  store_tokens_path   = var.cloudfoundry_store_tokens_path != "" ? var.cloudfoundry_store_tokens_path : null
  version             = "~> 0.12"
}
```

Pulling from a .tfvars file with the following set:

```
cloudfoundry_api_url = "https://api.your.cf.provider"
cloudfoundry_user = ""
cloudfoundry_password = ""
cloudfoundry_store_tokens_path = "/home/nick/work/cftokens/cftokens"
cloudfoundry_sso_passcode = "aBcdefghiJK2"
cloudfoundry_skip_ssl_validation = true
```

**Comment by janosbinder** [comment]
Hi @nickromney ,

So to fully understand, you use the provider with a `sso_pathcode` and with `user` and `password` at the same time? If yes, what is the usecase for that?

**Comment by nickromney** [comment]
Hi @janosbinder,

I use the provider with either:

- `user`, `password`
OR
- `sso_passcode`, `store_tokens_path`

This allows authentication for:
- a CI/CD pipeline with `user` and `password`
- colleagues with `user` and `password` (who've not yet migrated to SSO)
- users with SSO enabled (who in our implementation can no longer use user/password)

I've updated my comment above to show two possible `.tfvars` files

It was `store_tokens_path` which made the difference, when using with `sso_passcode`. In @AndreyZykov's [sample Git repo](https://github.com/AndreyZykov/cf-terraform-test/blob/master/main.tf) `store_tokens_path` isn't present/set.

**Pull request merged**